File: app/src/main/java/com/bzh/dytt/home/SingleListFragment.java
package com.bzh.dytt.home;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.Observer;
import android.arch.lifecycle.ViewModel;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import com.bzh.dytt.BaseFragment;
import com.bzh.dytt.R;
import com.bzh.dytt.data.network.Resource;

import java.util.List;

import butterknife.BindView;

public abstract class SingleListFragment<T> extends BaseFragment {

    private static final String TAG = "SingleListFragment";

    protected RecyclerView.Adapter mAdapter;

    protected ViewModel mViewModel;

    @BindView(R.id.swipe_refresh_layout)
    protected SwipeRefreshLayout mSwipeRefresh;

    @BindView(R.id.listview)
    protected RecyclerView mRecyclerView;
    protected Observer<Throwable> mThrowableObserver = new Observer<Throwable>() {
        @Override
        public void onChanged(@Nullable Throwable throwable) {
            if (getActivity() != null) {
                Toast.makeText(getActivity(), getResources().getString(R.string.fetch_video_detail_exception, throwable.getMessage()), Toast.LENGTH_SHORT).show();
            } else {
                Log.e(TAG, "onChanged: activity is null");
            }
        }
    };
    @BindView(R.id.empty_layout)
    View mEmpty;
    @BindView(R.id.error_layout)
    View mError;
    protected Observer<Resource<List<T>>> mListObserver = new Observer<Resource<List<T>>>() {
        @Override
        public void onChanged(@Nullable Resource<List<T>> result) {

            mEmpty.setVisibility(View.GONE);
            mError.setVisibility(View.GONE);

            assert result != null;
            switch (result.status) {
                case ERROR: {
                    mSwipeRefresh.setRefreshing(false);
                    mError.setVisibility(View.VISIBLE);
                }
                break;
                case LOADING: {
                    mSwipeRefresh.setRefreshing(true);
                }
                break;
                case SUCCESS: {
                    mSwipeRefresh.setRefreshing(false);
                    if (result.data == null || result.data.isEmpty()) {
                        mEmpty.setVisibility(View.VISIBLE);
                    } else {
                        replace(result.data);
                    }
                }
                break;
            }
        }
    };
    private SwipeRefreshLayout.OnRefreshListener mRefreshListener = new SwipeRefreshLayout.OnRefreshListener() {
        @Override
        public void onRefresh() {
            doRefresh();
        }
    };

    @Override
    protected View doCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        return inflater.inflate(R.layout.single_list_page, container, false);
    }

    protected abstract RecyclerView.Adapter createAdapter();

    protected abstract void replace(List<T> listData);

    protected abstract LiveData<Resource<List<T>>> getListLiveData();

    protected LiveData<Throwable> getThrowableLiveData() {
        return null;
    }

    protected abstract ViewModel createViewModel();

    @Override
    protected void doCreate(@Nullable Bundle savedInstanceState) {
        super.doCreate(savedInstanceState);
    }

    @Override
    protected void doViewCreated(View view, Bundle savedInstanceState) {
        super.doViewCreated(view, savedInstanceState);
        mViewModel = createViewModel();
        mSwipeRefresh.setOnRefreshListener(mRefreshListener);
        if (getListLiveData() != null) {
            getListLiveData().observe(this, getListObserver());
        }
        if (getThrowableLiveData() != null) {
            getThrowableLiveData().observe(this, getThrowableObserver());
        }
        mAdapter = createAdapter();
        mRecyclerView.setLayoutManager(new LinearLayoutManager(getActivity()));
        mRecyclerView.setAdapter(getAdapter());
    }

    protected void doRefresh() {
    }

    public Observer<Resource<List<T>>> getListObserver() {
        return mListObserver;
    }

    public RecyclerView.Adapter getAdapter() {
        return mAdapter;
    }

    public Observer<Throwable> getThrowableObserver() {
        return mThrowableObserver;
    }
}

File: app/src/main/java/com/bzh/dytt/SingleListFragment.kt
package com.bzh.dytt

import android.arch.lifecycle.LiveData
import android.arch.lifecycle.Observer
import android.arch.lifecycle.ViewModel
import android.os.Bundle
import android.support.v4.widget.SwipeRefreshLayout
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.support.v7.widget.RecyclerView.SCROLL_STATE_IDLE
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import com.bzh.dytt.data.ExceptionType
import com.bzh.dytt.data.Resource
import com.bzh.dytt.data.Status
import com.orhanobut.logger.Logger
import kotlinx.android.synthetic.main.error_layout.*
import kotlinx.android.synthetic.main.single_list_page.*


abstract class SingleListFragment<T> : BaseFragment() {

    protected var mOtherExceptionObserver: Observer<Resource<ExceptionType>> = Observer { result -> onOtherException(result) }

    protected var mRefreshListener: SwipeRefreshLayout.OnRefreshListener = SwipeRefreshLayout.OnRefreshListener { doRefresh() }

    protected var mListObserver: Observer<Resource<List<T>>> = Observer { result ->

        mEmpty.visibility = View.GONE
        mError.visibility = View.GONE

        when (result?.status) {

            Status.ERROR -> {
                mSwipeRefresh.isRefreshing = false
                mError.visibility = View.VISIBLE
            }

            Status.LOADING -> {
                mSwipeRefresh.isRefreshing = true
            }

            Status.SUCCESS -> {
                mSwipeRefresh.isRefreshing = false
                if (result.data == null || result.data.isEmpty()) {
                    mEmpty.visibility = View.VISIBLE
                } else {
                    if (IsScrollIdle()) {
                        replace(result.data)
                    }
                }
            }
        }
    }

    protected val mScrollListener = object : RecyclerView.OnScrollListener() {

        override fun onScrollStateChanged(recyclerView: RecyclerView?, newState: Int) {
            super.onScrollStateChanged(recyclerView, newState)
            mIsScrollIdle = newState == SCROLL_STATE_IDLE
        }

        override fun onScrolled(recyclerView: RecyclerView?, dx: Int, dy: Int) {
            super.onScrolled(recyclerView, dx, dy)
        }
    }

    protected lateinit var mAdapter: RecyclerView.Adapter<*>

    protected lateinit var mViewModel: ViewModel

    protected lateinit var mSwipeRefresh: SwipeRefreshLayout

    protected lateinit var mRecyclerView: RecyclerView

    protected lateinit var mEmpty: View

    protected lateinit var mError: View

    private var mIsScrollIdle = true

    override fun doCreate(savedInstanceState: Bundle?) {
        super.doCreate(savedInstanceState)
    }

    override fun doCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        return inflater.inflate(R.layout.single_list_page, container, false)
    }

    override fun doViewCreated(view: View, savedInstanceState: Bundle?) {
        super.doViewCreated(view, savedInstanceState)

        mSwipeRefresh = swipe_refresh_layout
        mRecyclerView = listview
        mEmpty = empty_layout
        mError = error_layout

        mViewModel = createViewModel()
        mSwipeRefresh.setOnRefreshListener(mRefreshListener)

        getListLiveData().observe(this, getListObserver())

        if (getThrowableLiveData() != null) {
            getThrowableLiveData()?.observe(this, getThrowableObserver())
        }

        mAdapter = createAdapter()
        mRecyclerView.layoutManager = LinearLayoutManager(activity)
        mRecyclerView.adapter = getAdapter()
        mRecyclerView.addOnScrollListener(mScrollListener)
    }

    protected open fun doRefresh() {

    }

    protected open fun getListObserver(): Observer<Resource<List<T>>> {
        return mListObserver
    }

    protected open fun onOtherException(result: Resource<ExceptionType>?) {
        if (result?.data == null) {
            return
        }

        when (result.data) {
            ExceptionType.TaskFailure -> {
                if (activity != null) {
                    Toast.makeText(activity, resources.getString(R.string.fetch_video_detail_exception, result.message), Toast.LENGTH_SHORT).show()
                } else {
                    Logger.e(TAG, "onChanged: activity is null")
                }
            }
        }
    }

    protected fun IsScrollIdle() = mIsScrollIdle

    protected open fun getThrowableLiveData(): LiveData<Resource<ExceptionType>>? {
        return null
    }

    protected fun getAdapter(): RecyclerView.Adapter<*> {
        return mAdapter
    }

    protected fun getThrowableObserver(): Observer<Resource<ExceptionType>> {
        return mOtherExceptionObserver
    }

    protected abstract fun createAdapter(): RecyclerView.Adapter<*>

    protected abstract fun replace(listData: List<T>)

    protected abstract fun getListLiveData(): LiveData<Resource<List<T>>>

    protected abstract fun createViewModel(): ViewModel

    companion object {
        private const val TAG = "SingleListFragment"
    }
}